gs_quant/api/fred/fredapi.py

```diff
@@ -0,0 +1,305 @@
+        """
+        
+        if end is not None:
+            assert(isinstance(end, dt.datetime) or isinstance(end, dt.date)), \
+                "End must be of type dt.datetime or dt.date!"
+        if start is not None:
+            assert(isinstance(start, dt.datetime) or isinstance(start, dt.date)), \
+                "Start must be of type dt.datetime or dt.date!"
+        if start is not None and end is not None:
+            if type(start) != type(end):
+                raise ValueError('Start and end types must match!')
+        
+        url = ''
+        date_dict = {
+            "observation_start" : start,
+            "observation_end" : end,
+            "realtime_end" : as_of,
+            "realtime_start" : since
+        }
+        for date in date_dict.keys():
+            if date_dict[date] is not None:
+                url += '&{}={}'.format(date, date_dict[date])
+
+        if kwargs.keys():
+            url += '&{}'.format(urlencode(kwargs))
+
+        final_url = '{}/series/observations?series_id={{}}&api_key={}{}&\
+            file_type=json'.format(self.root_url, self.api_key, url)
+        return final_url
+
+    def query_data(
+        self, 
+        query: str, 
+        id: str, 
+        asset_id_type=None
+    ) -> dict:
+        """
+        Query data given a valid FRED series id and url. 
+        Will raise an HTTPError if the response was an HTPP error.
+
+        :param query: A url string of the requested data
+        :param id: A FRED series id 
+        :return: with id as key and requested DataFrame as value.
+        """
+
+        final_query = query.format(id)
+        series_dict = {}
+        try: 
+            response = requests.get(final_query)
+            response.raise_for_status() 
+            json_data = response.json()
+        except HTTPError as e:
+            raise ValueError(response.json()["error_message"])
+        if not len(json_data["observations"]):
+            raise ValueError("No data exists for {} for the provided paramters... ".format(id))
+        
+        data = pd.DataFrame(json_data["observations"])[["date", "value"]]
+        data['date'] = pd.to_datetime(data['date'])
+        data = data.set_index('date')['value']
+        data = data.sort_index().groupby('date').tail(1)
+        series_dict[id] = data
+        return series_dict
+
+    def last_data(
+        self,
+        query: str, 
+        id: str, 
+        **kwargs
+    ) -> dict:
+        """
+        Get the last point for a data series, at or before as_of
+
+        :param query: A url string of the requested data
+        :param id: A FRED series id 
+        :param kwargs: Extra query arguments
+        :return: with id as key and last point for this DataSet as value.
+        """
+        
+        data = self.query_data(query, id)
+        last_data_chunk = {}
+        for id in data:
+            last_data_chunk[id] = data[id].last("1D")
+        return last_data_chunk
```

Comment:
This can be more simply written as return {id: datum.last('1D') for id, datum in data.items()}